### DISCUSSION TITLE
About the `dictionary`, `example` and `alternative` methods

### DISCUSSION THREAD

**Discussion by Animenosekai** [Animenosekai/translate#62] [Ideas]
The `dictionary` and `example` methods are now too unstable/variable to be used.

Indeed, the values returned varies from one translator to another.

I wonder if it shouldn't be better to modify those methods :

- `example` should output a list of examples to use the given word. I also wonder what we should do with translators also returning the position of the word in the resulting sentences (maybe we should make a special *“Result”* class which holds the data and having the attributes storing the position of the word as nullable)

- `dictionary` should output a list of definitions of the given word, as expected of a dictionary. It might also give some features (etymology, gender, etc.).

- `alternative` should be a new method which outputs a list of alternative translations. We should maybe add a [python *property*](https://docs.python.org/3/library/functions.html#property) to the `TranslationResult` model, which holds a `AlternativeResult` model if already given when translating by the translator, or which fetches alternatives if it doesn't.

**Comment by Animenosekai** [comment]
@ZhymabekRoman what do you think about this ?

**Comment by ZhymabekRoman** [comment]
I thought about implementing `example` and `dictionary` when I was refactoring code for v2, but I couldn't think of anything other than making a separate response model for each translator, so the idea stayed, in `ExampleResult` you can find a couple of references for the future. At the moment I have something like the following model in my head:

```python
class ExampleResultList:
    def __init__(self, service, source, source_language, destination_language, results):
        self.service = service
        self.source = str(source)
        self.source_language = Language(source_language)
        self.destination_language = Language(destination_language)
        self.results = results  # results = [ExampleResult(...), ExampleResult(...)]


class ExampleResult:
	source_result: # Hello, how are you
	destination_result: # Hola, ¿cómo estás?
	reference: # The source from which the example was taken. Can be None, because not all translators provide this parameter (for example Bing and Microsoft), and even if they do, they vary greatly (for example Reverso and Yandex). I think it would be better if each translator returned a own `ReferenceResponse` model. 

	# And i think 'refernce' model field can be renamed to 'source', but i am not sure

	# Supported by translate service, but not implemented by translatepy example method: DeepL, Google
```

At the moment I can't say anything definite about the dictionary, I haven't even started to look at what the translation services return, I'm afraid they will have to sweat to bring all the translators into a unified response model

> * `alternative` should be a new method which outputs a list of alternative translations. We should maybe add a [python _property_](https://docs.python.org/3/library/functions.html#property) to the `TranslationResult` model, which holds a `AlternativeResult` model if already given when translating by the translator, or which fetches alternatives if it doesn't.

It's great idea, really necessary and useful functionality, all translators support it. With Lazy-load mechanism using property would be awesome

By the way, someone decided to make a port of translatepy in C#, the logic of all translators is very similar to what is implemented
https://github.com/d4n3436/GTranslate

**Reply by Animenosekai** [reply]
> ```python
> class ExampleResultList:
>     def __init__(self, service, source, source_language, destination_language, results):
>         self.service = service
>         self.source = str(source)
>         self.source_language = Language(source_language)
>         self.destination_language = Language(destination_language)
>         self.results = results  # results = [ExampleResult(...), ExampleResult(...)]
> 
> 
> class ExampleResult:
> 	source_result: # Hello, how are you
> 	destination_result: # Hola, ¿cómo estás?
> 	reference: # The source from which the example was taken. Can be None, because not all translators provide this parameter (for example Bing and Microsoft), and even if they do, they vary greatly (for example Reverso and Yandex). I think it would be better if each translator returned a own `ReferenceResponse` model. 
> 
> 	# And i think 'refernce' model field can be renamed to 'source', but i am not sure
> 
> 	# Supported by translate service, but not implemented by translatepy example method: DeepL, Google
> ```

I think that the example method should return a list of examples. What is the `source_result` and the `destination_result`, which one holds the example ?

The reference attribute is nice and could be a nullable attribute.
We should maybe add the position of the word ? Although I'm not sure if it is necessary, and it could be done by the end user.

**Reply by Animenosekai** [reply]
> At the moment I can't say anything definite about the dictionary, I haven't even started to look at what the translation services return,  I'm afraid they will have to sweat to bring all the translators into a unified response model

About the `dictionary` method, I'm not sure if we have this kind of method already implemented or if we should try finding new endpoints, but it would essentially be a model with the definition/meaning of the given word or the different words in the sentence. It's kind of like when you double-click on a word in a translation on Google Translate.

> I haven't even started to look at what the translation services return

I think that now they all have different meaning of the “dictionary”, which we should unify in the most basic/instinctive meaning of the word “dictionary” we have, something which gives the meaning of a word.

**Reply by Animenosekai** [reply]
> By the way, someone decided to make a port of translatepy in C#, the logic of all translators is very similar to what is implemented
https://github.com/d4n3436/GTranslate

Wow, didn't know that. I'll definitely take a look at it.

**Reply by Animenosekai** [reply]
<img width="1438" alt="Capture d’écran 0004-06-26 à 21 51 15" src="https://user-images.githubusercontent.com/40539549/175831508-8430d0e1-82b6-4dca-b348-764ea028ec37.png">

For instance, this is the information returned on Google Translate when double-clicking "how"

**Reply by ZhymabekRoman** [reply]
> I think that the example method should return a list of examples.

Yes, I am of the same opinion, I tried to implement in Yandex - https://github.com/ZhymabekRoman/translate/commit/f20960b07120c86300eb3ba429749b0821777a3c

> What is the `source_result` and the `destination_result`, which one holds the example ?

For example, take that we want to know examples of the use of the word "Hello" in Japanese. Almost all services return an example of the use of this word in the context of the original English language and in the context of the translation in Japanese. In our case `source_result` will be "*Hello*, I am Pete Lavache from Platforms Marketing", and `destination_result` = "*こんにちは*、プラットフォーム・マーケッティングのPete　Lavacheです。". I hope I explained clearly, and did not overdo it


> We should maybe add the position of the word ? Although I'm not sure if it is necessary, and it could be done by the end user.

This is very necessary because a word in a example can be different what it appears. For example, the word "Hello" in the examples may be "Regards", "Hullo", "Hallo", etc. But I honestly don't know how to pass this data, just as an index of word's start and end positions? Or maybe should return word inside the `ExampleWordResult` (maybe need to find a better naming) model at once in example sentence? Very convenient, but I don't know how it would work in practice

**Reply by d4n3436** [reply]
> By the way, someone decided to make a port of translatepy in C#, the logic of all translators is very similar to what is implemented
https://github.com/d4n3436/GTranslate

@ZhymabekRoman It's a bit funny that you mention that, since I started the GTranslate project 2-3 years ago, before pytranslate existed (sources [here](https://github.com/d4n3436/Fergun/blob/f02f2cb3f232ec510b7a594fe4f1b8bbf9b3bc49/src/Fergun/APIs/Translators.cs) and [here](https://github.com/d4n3436/Fergun/blob/9abc8a3ba8dd06899a3c1d25e8958b92861578d5/src/APIs/GTranslate/GTranslator.cs)) before moving everything to its own repo. I did all the research and reverse-engineering by myself, except the Yandex.Translate API endpoint from Android and the Google Translate token generation.

Anyways, you guys have a pretty good lib here. Feel free to adapt any part you need from GTranslate into your lib (like the signature generation from the Microsoft Translator Android app, your lib probably needs it since Bing blocked its token generation endpoint).
